server/src/main/java/org/elasticsearch/common/io/stream/ByteArrayStreamInput.java

```diff
@@ -84,7 +84,10 @@ protected void ensureCanReadBytes(int length) throws EOFException {
     }
 
     @Override
-    public byte readByte() {
+    public byte readByte() throws IOException {
+        if (limit - pos <= 0) {
+            throw new EOFException();
```

Comment:
There is an existing helper method to provide a nice error message, see `StreamInput.throwEOF`.